---
title: "FGDvNGD analysis"
author: "Samarth Mathur"
date: "08/08/2023"
output:
  pdf_document:
    fig_caption: yes
    keep_tex: yes
  header-includes: \usepackage{float}
  html_document:
    df_print: paged
---


\flushleft

```{r settings, include = FALSE}
knitr::opts_knit$set(root.dir = "~/Documents/Postdoc_research/WGR/Results/April23") # Change directory accordingly
library(plyr)
library(dplyr)
library(ggplot2)
library(gridExtra)
library(ggpubr)
library(broom)
```

### _Estimating function genetic diversity (FGD)_ ###

From the $DoS$ calculations, we have the genes putatively under strong positive selection (upper 10% of the $DoS$ distribution) and the genes under purifying selection (lower 10% of the $DoS$ distribution).  We also included loss-of-function sites as sites potentially also under purifying selection.

$$ DoS =  \frac{D_n}{D_n +D_s} - \frac{P_n}{P_n + P_s}$$

```{r get_sites, message=FALSE, warning=FALSE, echo=FALSE}
setwd("~/Documents/Postdoc_research/WGR/Results/April23")
sites <- c("lof.lower10","upper10")
for (s in sites)
{
  a <- read.table(paste("sites/",s,".sites.csv",sep=""),header = T)
  name <- paste(s,".sites",sep="")
  assign(name,a)
}

snps <- NULL
for (s in sites)
{
  df <- data.frame(Type=s,SNPs=nrow(get(paste(s,".sites",sep=""))))
  snps <- rbind(snps,df)
}
snps[,1] <- c("Deleterious", "Adaptive")
print(snps)
```


```{r get_pops, message=FALSE, warning=FALSE, include=FALSE}
poplist <- read.csv("data/DataS1.csv", header = T)
poplist <- poplist[,c(1,3)]
pops <- sort(unique(poplist$Location))
pops <- pops[-c(4,13)] # Remove GRLL and STER
```

### Mutation load ###

```{r Load, include=TRUE}
df.Ndel <- NULL
df.Load <- NULL

for (pop in pops)
{
  freq <- read.table(paste("freq/",pop,".lof.lower10.freq",sep=""),header = F)
  ndel <- length(which(freq$V7 != 0))
  fdel <- length(which(freq$V7 >= 0.9))
  df <- data.frame(Pop=pop,Ndel=ndel,Nfreq=fdel)
  df.Ndel <- rbind(df.Ndel,df)
}

for (pop in pops)
{
  geno <- read.table(paste("geno/",pop,".lof.lower10.GT.FORMAT",sep=""), header = T)
  inds <- colnames(geno)[-c(1,2)]
  N <- ncol(geno)-2
  L <- nrow(geno)
  for (j in 1:N) # per-ind genotypes
  {
    gt <- geno[,(j+2)]
    het <- length(which(gt == "0/1")) + length(which(gt == "0|1"))+
           length(which(gt == "1/0")) + length(which(gt == "1|0"))
    
    hom <- length(which(gt == "1/1")) + length(which(gt == "1|1"))
    
    df <- data.frame(SampleID=inds[j],Pop=pop,Het=het, Hom=hom)
      df.Load <- rbind(df.Load,df)
  }
}

# Population mean
df.Load.pop <- ddply(df.Load, "Pop", summarise, meanHet=mean(Het),sdHet=sd(Het), meanHom=mean(Hom),sdHom=sd(Hom))

df.Load.pop <- inner_join(df.Ndel,df.Load.pop, by="Pop")

#Write results
write.csv(df.Load,"results/df.Load.csv",quote = F, row.names = F)
write.csv(df.Load.pop,"results/df.Load.pop.csv",quote = F, row.names = F)
```


### $\pi_{adaptive}$ ###

```{r Piupper10, include=TRUE}

cols <- c("Region","Chr","Center","tW","tP","tF","tH","tL","Tajima", "fuf","fud","fayh","zeng","nSites")

df.pi.u10 <- NULL

for (pop in pops)
{
  pi_site <- read.table(paste("pi/upper10/",pop,".upper10_sites.pestPG",sep=""), header = F)
  colnames(pi_site) <- cols
  pi_site <- pi_site[-which(pi_site$nSites == 0),]
  pi_site <- pi_site[,c(2,5)]
  
  pi_cds <- read.table(paste("pi/upper10/",pop,".upper10_justCDS.pestPG",sep=""), header = F)
  colnames(pi_cds) <- cols
  pi_cds <- pi_cds[,c(2,14)]
  pi <- inner_join(pi_site,pi_cds,by="Chr")
  df <- data.frame(Pop=pop,pi_u10=mean(pi$tP/pi$nSites),sd_pi_u10=sd(pi$tP/pi$nSites))
  df.pi.u10 <- rbind(df.pi.u10,df)
}
#Write results
write.csv(df.pi.u10,"results/df.pi.upper10.csv",quote = F, row.names = F) 

```

### Results ###


```{r, echo=FALSE}

colnames(df.Load.pop)[3:7] <- c("DriftLoad","meanLinbreed","sdLinbreed","meanLrealized","sdLreallized")

fgd <- inner_join(df.Load.pop,df.pi.u10,by="Pop")

write.csv(fgd,"results/FGD.pop.csv",quote = F, row.names = F)
knitr::kable(fgd, align = "cccccc", caption = "mean and SD for different components of FGD", format.args = list(big.mark = ",", scientific = FALSE),digits = c(0,0,0,0,0,0,0,6,6))
```


### _Estimating neutral genetic diversity (NGD) at different time scales_ ###

#### _Long-term_ ####

```{r get_geneDesPi, include=FALSE, echo=TRUE}

df.geneDesPi <- NULL
for (pop in pops)
{
  pi <- read.table(paste("pi/geneDes/",pop,".outGene500_10_5.pestPG",sep=""), header = F)
  pi <- pi[which(pi$V14 !=0),]
  pi <- cbind(pi,pi$V5/pi$V14)
  colnames(pi)[15] <- "PI"
  df <- data.frame(Pop=pop,meanPi=mean(pi$PI),sdPi=sd(pi$PI))
  df.geneDesPi <- rbind(df.geneDesPi,df)
}

write.csv(df.geneDesPi,"results/df.geneDesPi.csv",quote = F, row.names = F)

```


#### _Medium-term_ ####

$$ F_{ROH,t} =  1 - (1-\frac{1}{2N_e})^t$$
which means,
$$ N_e = \frac{1}{2[1-(1-F_{ROH,t})^\frac{1}{t}]}$$
where $F_{ROH,t}$ corresponds to the average population inbreeding coefficient for a given time period and $t$ is the maximum nuumber of generations in that time period.\
Thus, in our case $F_{ROH,t}$ is $F_{ROH}$ in 500-700kb length and $t$ is 35 generations.

```{r get_FROH, include=FALSE}
roh <- read.table("data/onlyScat.10x.wholegenome.hom", header = T)
colnames(roh)[1] <- "SampleID"
roh <- inner_join(roh,poplist, by="SampleID")
roh <- roh[order(roh$SampleID),]

# ROH between 500-700kb
roh_57 <- roh[which(roh$KB >= 500 & roh$KB <= 700),]
roh_indv <- ddply(roh_57, "SampleID", summarise, sumROH=sum(KB))
roh_indv <- inner_join(roh_indv,poplist, by="SampleID")

seq_len <- 1346095645 # Ref genome length (No Z chr and scaffs > 500kb)
roh_indv <- cbind(roh_indv,roh_indv$sumROH*1000/seq_len)
colnames(roh_indv)[4] <- "FROH"
roh_indv <- roh_indv[-which(roh_indv$Location == "GRLL"),]
colnames(roh_indv)[3] <- "Pop" 

load.indv <- inner_join(df.Load,roh_indv,by="SampleID")

froh.mean <- ddply(roh_indv, "Pop", summarise, meanFROH=mean(FROH), sdFROH=sd(FROH))

# froh = 1-(1-1/2ne)^t
deno_mean <- 2*(1-(1-froh.mean$meanFROH)^(1/35))
ne_mean <- 1/deno_mean

froh.mean <- cbind(froh.mean,ne_mean)
colnames(froh.mean)[4] <- c("NeROH")
write.csv(froh.mean,"results/meanFROH.csv",quote = F, row.names = F)
```

#### _Short-term_ ####

```{r get_neLd, include=FALSE, echo=FALSE}

neld_runs <- read.csv("data/allPop.LDNe.csv", header = T)
neld.pop <- ddply(neld_runs, "Pop", summarise, meanNeLD=mean(NeLD), sdNeLD=sd(NeLD))

write.csv(neld.pop,"results/meanNeLD.csv",quote = F, row.names = F)
```

### Results ###

```{r, echo=FALSE}
ngd <- inner_join(df.geneDesPi,froh.mean,by="Pop")
ngd <- inner_join(ngd,neld.pop,by="Pop")

write.csv(ngd,"results/NGD.pop.csv",quote = F, row.names = F)
knitr::kable(ngd, align = "cccccc", caption = "mean and SD for different components of NGD", format.args = list(big.mark = ",", scientific = FALSE))
```

### _FGDvNGD at different time scales_ ###

```{r fgdvngd, include=FALSE}

fgdvngd <- inner_join(fgd,ngd,by="Pop")
write.csv(fgdvngd,"results/FGDvNGD.pop.csv",quote = F, row.names = F)

fgdvngd <- read.csv("results/FGDvNGD.pop.csv",header = T)
# Long term
p1 <- ggplot(fgdvngd,aes(fill=Pop))+ # Ndel vs Pi
  geom_point(mapping=aes(y=Ndel, x=meanPi), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=Ndel, x=meanPi), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$meanPi, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$Ndel, n = 5))+
  labs(x=expression(paste("Nucleotide diversity (",bar(pi)[geneDes], ")")),
       y=expression(paste("No. of deleterious mutations (",N[del], ")")))

p2 <- ggplot(fgdvngd,aes(fill=Pop))+ # DriftLoad vs Pi
  geom_point(mapping=aes(y=DriftLoad, x=meanPi), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=DriftLoad, x=meanPi), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$meanPi, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$DriftLoad, n = 5))+
  labs(x=expression(paste("Nucleotide diversity (",bar(pi)[geneDes], ")")),
       y=expression(paste("Drift Load (",L[drift], ")")))

p3 <- ggplot(fgdvngd,aes(fill=Pop))+ # meanLinbreed vs Pi
  geom_point(mapping=aes(y=meanLinbreed, x=meanPi), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=meanLinbreed, x=meanPi), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$meanPi, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$meanLinbreed, n = 5))+
  labs(x=expression(paste("Nucleotide diversity (",bar(pi)[geneDes], ")")),
       y=expression(paste("Inbreeding load (",bar(L)[inbreeding], ")")))

p4 <- ggplot(fgdvngd,aes(fill=Pop))+ # meanLrealized vs Pi
  geom_point(mapping=aes(y=meanLrealized, x=meanPi), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=meanLrealized, x=meanPi), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$meanPi, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$meanLrealized, n = 5))+
  labs(x=expression(paste("Nucleotide diversity (",bar(pi)[geneDes], ")")),
       y=expression(paste("Realized load (",bar(L)[realized], ")")))

p5 <- ggplot(fgdvngd,aes(fill=Pop))+ # pi_u10 vs Pi
  geom_point(mapping=aes(y=pi_u10, x=meanPi), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=pi_u10, x=meanPi), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$meanPi, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$pi_u10, n = 5))+
  labs(x=expression(paste("Nucleotide diversity (",bar(pi)[geneDes], ")")),
       y=expression(paste("Upper10% nucleotide diversity (",pi[upper10], ")")))


# Medium term  (NeROH)

p6 <- ggplot(fgdvngd,aes(fill=Pop))+ # Ndel vs NeROH
  geom_point(mapping=aes(y=Ndel, x=NeROH), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=Ndel, x=NeROH), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$NeROH, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$Ndel, n = 5))+
  labs(x=expression(paste(Ne[ROH])),
       y=expression(paste("No. of deleterious mutations (",N[del], ")")))

p7 <- ggplot(fgdvngd,aes(fill=Pop))+ # DriftLoad vs NeROH
  geom_point(mapping=aes(y=DriftLoad, x=NeROH), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=DriftLoad, x=NeROH), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$NeROH, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$DriftLoad, n = 5))+
  labs(x=expression(paste(Ne[ROH])),
       y=expression(paste("Drift Load (",L[drift], ")")))


p8 <- ggplot(fgdvngd,aes(fill=Pop))+ # meanLinbreed vs NeROH
  geom_point(mapping=aes(y=meanLinbreed, x=NeROH), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=meanLinbreed, x=NeROH), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$NeROH, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$meanLinbreed, n = 5))+
  labs(x=expression(paste(Ne[ROH])),
       y=expression(paste("Inbreeding load (",bar(L)[inbreeding], ")")))

p9 <- ggplot(fgdvngd,aes(fill=Pop))+ # meanLrealized vs NeROH
  geom_point(mapping=aes(y=meanLrealized, x=NeROH), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=meanLrealized, x=NeROH), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$NeROH, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$meanLrealized, n = 5))+
  labs(x=expression(paste(Ne[ROH])),
       y=expression(paste("Realized load (",bar(L)[realized], ")")))

p10 <- ggplot(fgdvngd,aes(fill=Pop))+ # pi_u10 vs NeROH
  geom_point(mapping=aes(y=pi_u10, x=NeROH), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=pi_u10, x=NeROH), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$NeROH, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$pi_u10, n = 5))+
  labs(x=expression(paste(Ne[ROH])),
       y=expression(paste("Upper10% nucleotide diversity (",pi[upper10], ")")))


# Medium term  (meanFROH)

p6b <- ggplot(fgdvngd,aes(fill=Pop))+ # Ndel vs meanFROH
  geom_point(mapping=aes(y=Ndel, x=meanFROH), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=Ndel, x=meanFROH), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$meanFROH, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$Ndel, n = 5))+
  labs(x=expression(paste(bar(F)[ROH])),
       y=expression(paste("No. of deleterious mutations (",N[del], ")")))

p7b <- ggplot(fgdvngd,aes(fill=Pop))+ # DriftLoad vs meanFROH
  geom_point(mapping=aes(y=DriftLoad, x=meanFROH), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=DriftLoad, x=meanFROH), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$meanFROH, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$DriftLoad, n = 5))+
  labs(x=expression(paste(bar(F)[ROH])),
       y=expression(paste("Drift Load (",L[drift], ")")))


p8b <- ggplot(fgdvngd,aes(fill=Pop))+ # meanLinbreed vs meanFROH
  geom_point(mapping=aes(y=meanLinbreed, x=meanFROH), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=meanLinbreed, x=meanFROH), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$meanFROH, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$meanLinbreed, n = 5))+
  labs(x=expression(paste(bar(F)[ROH])),
       y=expression(paste("Inbreeding load (",bar(L)[inbreeding], ")")))

p9b <- ggplot(fgdvngd,aes(fill=Pop))+ # meanLrealized vs meanFROH
  geom_point(mapping=aes(y=meanLrealized, x=meanFROH), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=meanLrealized, x=meanFROH), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$meanFROH, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$meanLrealized, n = 5))+
  labs(x=expression(paste(bar(F)[ROH])),
       y=expression(paste("Realized load (",bar(L)[realized], ")")))

p10b <- ggplot(fgdvngd,aes(fill=Pop))+ # pi_u10 vs meanFROH
  geom_point(mapping=aes(y=pi_u10, x=meanFROH), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=pi_u10, x=meanFROH), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$meanFROH, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$pi_u10, n = 5))+
  labs(x=expression(paste(bar(F)[ROH])),
       y=expression(paste("Upper10% nucleotide diversity (",pi[upper10], ")")))

# Short Term

p11 <- ggplot(fgdvngd,aes(fill=Pop))+ # Ndel vs meanNeLD
  geom_point(mapping=aes(y=Ndel, x=meanNeLD), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=Ndel, x=meanNeLD), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$meanNeLD, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$Ndel, n = 5))+
  labs(x=expression(paste(bar(Ne)[LD])),
       y=expression(paste("No. of deleterious mutations (",N[del], ")")))

p12 <- ggplot(fgdvngd,aes(fill=Pop))+ # DriftLoad vs meanNeLD
  geom_point(mapping=aes(y=DriftLoad, x=meanNeLD), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=DriftLoad, x=meanNeLD), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$meanNeLD, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$DriftLoad, n = 5))+
  labs(x=expression(paste(bar(Ne)[LD])),
       y=expression(paste("Drift Load (",L[drift], ")")))

p13 <- ggplot(fgdvngd,aes(fill=Pop))+ # meanLinbreed vs meanNeLD
  geom_point(mapping=aes(y=meanLinbreed, x=meanNeLD), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=meanLinbreed, x=meanNeLD), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$meanNeLD, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$meanLinbreed, n = 5))+
  labs(x=expression(paste(bar(Ne)[LD])),
       y=expression(paste("Inbreeding load (",bar(L)[inbreeding], ")")))

p14 <- ggplot(fgdvngd,aes(fill=Pop))+ # meanLrealized vs meanNeLD
  geom_point(mapping=aes(y=meanLrealized, x=meanNeLD), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=meanLrealized, x=meanNeLD), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$meanNeLD, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$meanLrealized, n = 5))+
  labs(x=expression(paste(bar(Ne)[LD])),
       y=expression(paste("Realized load (",bar(L)[realized], ")")))

p15 <- ggplot(fgdvngd,aes(fill=Pop))+ # pi_u10 vs meanNeLD
  geom_point(mapping=aes(y=pi_u10, x=meanNeLD), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=pi_u10, x=meanNeLD), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$meanNeLD, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$pi_u10, n = 5))+
  labs(x=expression(paste(bar(Ne)[LD])),
       y=expression(paste("Upper10% nucleotide diversity (",pi[upper10], ")")))

```

#### Long-term ####

```{r , warning=FALSE, echo=FALSE, message=FALSE, fig.cap=c("Relationship between different components of functional genetic diversity within population estimated at a per-site basis vs. nucleotide diversity "), fig.pos="H", fig.width=16, fig.height=10}

# Freq based
ggarrange(p1, p2, ncol=2, nrow=1, common.legend = TRUE, legend="bottom")

#Ind based
ggarrange(p3, p4,p5, ncol=3, nrow=1, common.legend = TRUE, legend="bottom")


```

#### Medium-term ####

```{r , warning=FALSE, echo=FALSE, message=FALSE, fig.cap=c("Relationship between different components of functional genetic diversity  within population estimated at a per-site basis vs. mean FROH"), fig.pos="H", fig.width=16, fig.height=10}

# Freq based
ggarrange(p6, p7, ncol=2, nrow=1, common.legend = TRUE, legend="bottom")
ggarrange(p6b, p7b, ncol=2, nrow=1, common.legend = TRUE, legend="bottom")

#Ind based
ggarrange(p8, p9,p10, ncol=3, nrow=1, common.legend = TRUE, legend="bottom")
ggarrange(p8b, p9b,p10b, ncol=3, nrow=1, common.legend = TRUE, legend="bottom")


```


#### Short-term ####

```{r , warning=FALSE, echo=FALSE, message=FALSE, fig.cap=c("Relationship between different components of functional genetic diversity  within population estimated at a per-site basis vs. mean NeLD"), fig.pos="H", fig.width=16, fig.height=10}

# Freq based
ggarrange(p11, p12, ncol=2, nrow=1, common.legend = TRUE, legend="bottom")

#Ind based
ggarrange(p13, p14,p15, ncol=3, nrow=1, common.legend = TRUE, legend="bottom")



```

### Statistics ###
```{r, echo=FALSE}
# Long-term
summary(lm(meanNdel ~ meanPi, data=fgdvngd)) # Adjusted R2 = 0.9098; p-value: 9.449e-07
summary(lm(meanLdrift ~ meanPi, data=fgdvngd)) # Adjusted R2 = 0.6778; p-value: 0.0001987

summary(lm(meanHet ~ meanPi, data=fgdvngd)) # Adjusted R2 = 0.6631; p-value: 0.0007699
summary(lm(meanHom ~ meanPi, data=fgdvngd)) # Adjusted R2 = 0.1547; p-value: 0.1133
summary(lm(pi_u10 ~ meanPi, data=fgdvngd)) # Adjusted R2 = 0.9384; p-value: 1.383e-07

# Medium-term
summary(lm(meanNdel ~ NeROH, data=fgdvngd)) # R2 = 0.07042; p-value: 0.4045
summary(lm(meanLdrift ~ NeROH, data=fgdvngd)) # R2 = 0.0854; p-value: 0.185

summary(lm(meanHet ~ NeROH, data=fgdvngd)) # R2 = 0.181; p-value: 0.09368
summary(lm(meanHom ~ NeROH, data=fgdvngd)) # R2 = 0.1183; p-value: 0.1466
summary(lm(pi_u10 ~ NeROH, data=fgdvngd)) # R2 = 0.03963; p-value: 0.2556


summary(lm(meanNdel ~ meanFROH, data=fgdvngd)) # R2 = 0.1582; p-value: 0.1104
summary(lm(meanLdrift ~ meanFROH, data=fgdvngd)) # R2 = 0.3065; p-value: 0.036

summary(lm(meanHet ~ meanFROH, data=fgdvngd)) # R2 = 0.4421; p-value: 0.01092
summary(lm(meanHom ~ meanFROH, data=fgdvngd)) # R2 = 0.3164; p-value: 0.03321
summary(lm(pi_u10 ~ meanFROH, data=fgdvngd)) # R2 = 0.2595; p-value: 0.05215

#Short-term
summary(lm(meanNdel ~ meanNeLD, data=fgdvngd)) # Adjusted R2 = 0.3523; p-value: 0.02462
summary(lm(meanLdrift ~ meanNeLD, data=fgdvngd)) # Adjusted R2 = 0.2693; p-value: 0.04832

summary(lm(meanHet ~ meanNeLD, data=fgdvngd)) # Adjusted R2 = 0.2867; p-value: 0.04217
summary(lm(meanHom ~ meanNeLD, data=fgdvngd)) # Adjusted R2 = 0.3177; p-value: 0.03285
summary(lm(pi_u10 ~ meanNeLD, data=fgdvngd)) # Adjusted R2 = 0.3805; p-value: 0.01928


```

### Correlations between different components of FGD ###

```{r FGD_corr, include=FALSE}

p16 <- ggplot(fgdvngd,aes(fill=Pop))+ # DriftLoad vs Ndel
  geom_point(mapping=aes(y=DriftLoad, x=Ndel), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=DriftLoad, x=Ndel), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$Ndel, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$DriftLoad, n = 5))+
  labs(x=expression(paste(N[del])),
       y=expression(paste("Drift load (",L[drift], ")")))

p17 <- ggplot(fgdvngd,aes(fill=Pop))+ # meanLrealized vs meanLinbreed
  geom_point(mapping=aes(y=meanLrealized, x=meanLinbreed), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=meanLrealized, x=meanLinbreed), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_y_continuous(breaks = pretty(fgdvngd$meanLrealized, n = 5))+
  scale_x_continuous(breaks = pretty(fgdvngd$meanLinbreed, n = 5))+
  labs(x=expression(paste("Inbreeding load (",L[inbreed], ")")),
       y=expression(paste("Realized load (",L[realized], ")")))

p18 <- ggplot(fgdvngd,aes(fill=Pop))+ # meanLrealized vs DriftLoad
  geom_point(mapping=aes(y=meanLrealized, x=DriftLoad), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=meanLrealized, x=DriftLoad), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_y_continuous(breaks = pretty(fgdvngd$meanLrealized, n = 5))+
  scale_x_continuous(breaks = pretty(fgdvngd$DriftLoad, n = 5))+
  labs(x=expression(paste("Drift load (",L[drift], ")")),
       y=expression(paste("Realized load (",L[realized], ")")))

p19 <- ggplot(fgdvngd,aes(fill=Pop))+ # pi_u10 vs meanLinbreed
  geom_point(mapping=aes(y=pi_u10, x=meanLinbreed), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=pi_u10, x=meanLinbreed), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_y_continuous(breaks = pretty(fgdvngd$pi_u10, n = 5))+
  scale_x_continuous(breaks = pretty(fgdvngd$meanLinbreed, n = 5))+
  labs(x=expression(paste("Inbreeding load (",L[inbreed], ")")),
       y=expression(paste("Upper10% nucleotide diversity (",pi[upper10], ")")))

```

```{r , warning=FALSE, echo=FALSE, message=FALSE, fig.cap=c("Relationship between different components of functional genetic diversity  within population "), fig.pos="H", fig.width=16, fig.height=10}

ggarrange(p16,p17, p18,p19, ncol=2, nrow=2, common.legend = TRUE, legend="bottom")
```

```{r, echo=FALSE}
summary(lm(DriftLoad ~ Ndel, data=fgdvngd)) # R2 = 0.7336; p-value: 0.0002296
summary(lm(meanLrealized ~ meanLinbreed, data=fgdvngd)) # R2 = 0.5644; p-value: 0.002936
summary(lm(meanLrealized ~ DriftLoad, data=fgdvngd)) # R2 = 0.1674; p-value: 0.1034
summary(lm(pi_u10 ~ meanLinbreed, data=fgdvngd)) # R2 = 0.6546; p-value: 0.0008757

```


Correlations between different estimates of Ne

```{r, echo=FALSE}

p20 <- ggplot(fgdvngd,aes(fill=Pop))+ # meanFROH vs meanPi
  geom_point(mapping=aes(y=meanFROH, x=meanPi), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=meanFROH, x=meanPi), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$meanPi, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$meanFROH, n = 5))+
  labs(x=expression(paste("Nucleotide diversity (",bar(pi)[geneDes], ")")),
       y=expression(paste(bar(F)[ROH])))

p21 <- ggplot(fgdvngd,aes(fill=Pop))+ # NeROH vs meanPi
  geom_point(mapping=aes(y=NeROH, x=meanPi), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=NeROH, x=meanPi), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$meanPi, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$NeROH, n = 5))+
  labs(x=expression(paste("Nucleotide diversity (",bar(pi)[geneDes], ")")),
       y=expression(paste(Ne[ROH])))

p22 <- ggplot(fgdvngd,aes(fill=Pop))+ # meanNeLD vs meanPi
  geom_point(mapping=aes(y=meanNeLD, x=meanPi), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=meanNeLD, x=meanPi), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$meanPi, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$meanNeLD, n = 5))+
  labs(x=expression(paste("Nucleotide diversity (",bar(pi)[geneDes], ")")),
       y=expression(paste(bar(Ne)[LD])))

p23 <- ggplot(fgdvngd,aes(fill=Pop))+ # meanNeLD vs meanFROH
  geom_point(mapping=aes(y=meanNeLD, x=meanFROH), color="black", size=5,  shape=21)+
  geom_smooth(method=lm , aes(y=meanNeLD, x=meanFROH), color="black", fill=adjustcolor("#69b3a2",alpha.f = 0.1), se=T) +
  theme_classic(base_size = 15)+
  scale_x_continuous(breaks = pretty(fgdvngd$meanFROH, n = 5))+
  scale_y_continuous(breaks = pretty(fgdvngd$meanNeLD, n = 5))+
  labs(x=expression(paste(bar(F)[ROH])),
       y=expression(paste(bar(Ne)[LD])))

ggarrange(p20,p21,p22,p23, ncol=2, nrow=2, common.legend = TRUE, legend="bottom")

summary(lm(meanFROH ~ meanPi, data=fgdvngd)) # Adjusted R-squared:  0.2873; p-value: 0.04197
summary(lm(NeROH ~ meanPi, data=fgdvngd)) # Adjusted R-squared:  0.07337; p-value: 0.2013
summary(lm(meanNeLD ~ meanPi, data=fgdvngd)) # Adjusted R-squared:  0.3414; p-value: 0.02701
summary(lm(meanNeLD ~ meanFROH, data=fgdvngd)) # Adjusted R-squared:  0.01637; p-value: 0.3023
```